cli/command/image/build.go

```diff
@@ -247,6 +253,66 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		buildCtx = replaceDockerfileTarWrapper(ctx, buildCtx, relDockerfile, translator, &resolvedTags)
 	}
 
+	// write ctx tar to disk, to make it readable for Image Reference extraction
+	originCtx := buildCtx
+	defer originCtx.Close()
+	readbackTar, err := ioutil.TempFile("", "docker-ctx")
```

Comment:
What about letting the CLI code pass in the contents of the Dockerfile to `runBuild` in the cases where the CLI can read the Dockerfile directly (this should be true in the vast majority of cases)? It might make the code a little more complicated, but I think avoiding an extra layer of buffering to disk would be a big win. With the current design in this PR, we wouldn't even be able to start sending the build context to the daemon until it had been fully buffered to disk and read back, so I think this would be a very noticeable performance regression. I see now that there are cases where this might be necessary, but in most cases the CLI is reading the `Dockerfile` from a file on disk and putting it inside the tar stream in the first place. Writing the whole tar stream to disk just to extract a file that we added in the first place would be quite inefficient.